react-native-hub-app/src/checklist.tsx

```diff
@@ -0,0 +1,482 @@
+    }
+  }
+
+  async getUserToken(database, id): Promise<string> {
+    let token;// = await AsyncStorage.getItem(TOKEN_KEY);
+    if (token) {
+      return token;
+    }
+    token = await database.getToken(id);
+    await AsyncStorage.setItem(TOKEN_KEY, token);
+    return token;
+  }
+
+  async getContext(): Promise<Context> {
+    // Pull the stored context to reuse if available && valid date
+    let contextStr = await AsyncStorage.getItem(CONTEXT_KEY);
+    if (contextStr) {
+      const ctxJson = JSON.parse(contextStr);
+      if (
+        ctxJson['x-textile-api-sig-msg'] && (Date.parse(ctxJson['x-textile-api-sig-msg'])) > (new Date()).getTime()) {
+        // Not expired
+        const ctx = Context.fromJSON(ctxJson, API_URL);
+        return ctx;
+      }
+    }
+  }
+
+  async getUserThread(db: Client, ctx: Context): Promise<ThreadID> {
+    let idStr = await AsyncStorage.getItem(USER_THREAD_ID);
+    if (idStr) {
+      /**
+       * Temporary hack to get ThreadID working in RN
+       */
+      const temp: ThreadID = ThreadID.fromString(idStr);
+      // @ts-ignore
+      const id = ThreadID.fromBytes(Buffer.from(temp.buf));
+      return id;
+    } else {
+      const id: ThreadID = ThreadID.fromRandom();
+      await AsyncStorage.setItem(USER_THREAD_ID, id.toString());
+
+      /**
+       * Each new ThreadID requires a `newDB` call.
+       */
+      await db.newDB(id, ctx.withThread(id))
+
+      /** 
+       * We add our first Collection to the DB for Astronauts.
+       */
+      await db.newCollection(id, 'Astronaut', astronautSchema);
+      return id;
+    }
+  }
+
+  async runStep(stepNumber: number) {
+    const steps = this.state.steps;
+    const data = steps[stepNumber];
+    try {
+      let {ctx, db, threadId} = this.state;
+      switch (stepNumber) {
+        case 0: {
+          let updatedState = {};
+          /**
+           * Create a new user Identity
+           * 
+           * The identity pk will be cached in AsyncStorage.
+           * On the next session, the pk identity will be reused
+           */
+          const id = await this.generateIdentity();
+          
+          /**
+           * Context contains the token and session information
+           * 
+           * If possible, we'll reuse an existing session. 
+           * If it doesn't exist or is expired, we'll create a new one.
+           */
+          const existingCtx = await this.getContext();
+          if (existingCtx) {
+            db = new Client(existingCtx);
+            ctx = existingCtx;
+            data.message = 'Using existing Identity'
+          } else {
+            /** 
+             * Create a new Context (API_URL can be blank)
+             */
+            ctx = new Context(API_URL);
+
+            /**
+             * Create a new Signature from your key secret
+             * 
+             * note: this should be done offline
+             */
+            const sig = await createAPISig(USER_API_SECRET);
```

Comment:
It feels like the `sig` is what we should really be embedding in an app. so the dev would generate a sig one time for all users, that should expire in X time or a year or something. not sure. or maybe this is just the right way.

Reply:
Note that this whole bit and the next line can be wrapped up in this single call:

```
ctx.withUserKey(USER_API_KEY) // Assuming this is { key: string, secret: string, type: 1 }
```

Note also that you no longer have to re-assign to `ctx` because it is mutable (I'm still not sure I like that it is mutable, but that's what we landed on for now so let's see how it feels).

Reply:
hmm... i wonder what happens when i mutate a state variable like that in a component. will the mutated state really be persisted between cycles...

Reply:
It shouldn't be any diff than what you're doing now? I just mean you don't have to do:

```
ctx = ctx.With*
```

But you'd likely still have to call setState if it changes I guess?

Reply:
what is type?

Reply:
oh it just comes from the API, its the Key type (0 = ACCOUNT, 1 = USER), it actually isn't required here.